### PR TITLE
fix: Merge AddressSearch and FieldAddress components

### PR DESCRIPTION
**Pull request by rafenden**
## Before

image

## After

image

### CODE REVIEW

**Comment by web-bert** [comment]
Why do we want the postcode field to be so wide?

**Comment by rafenden** [comment]
> Why do we want the postcode field to be so wide?
The size of this field was decreased, see screenshot.

File: src/address-search/usePostcodeLookup.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    const addressCount = createAddressCount(data)
+
+    return [
+      addressCount,
+      ...data,
+    ]
+  }
+
+  return {
+    createAddressCount,
```
Comment (yeahfro) [inline src/address-search/usePostcodeLookup.js]: It doesn't look like you need to effectively export this as it's a private function.

File: src/forms/elements/FieldAddress.jsx [hunk 2]
```diff
@@ -1,31 +1,54 @@
...
+import usePostcodeLookup from '../../address-search/usePostcodeLookup'
 import FieldInput from './FieldInput'
 import FieldUneditable from './FieldUneditable'
+import FieldWrapper from './FieldWrapper'
+import StatusMessage from '../../status-message/StatusMessage'
 
 const UNITED_KINGDOM = 'United Kingdom'
 
+const StyledFieldPostcode = styled(FieldInput)`
+  ${MEDIA_QUERIES.TABLET} {
```
Comment (yeahfro) [inline src/forms/elements/FieldAddress.jsx]: Looks like there's a curly brace missing here

File: src/forms/elements/FieldAddress.jsx [hunk 3]
```diff
@@ -1,31 +1,54 @@
...
-
-  const { setFieldValue } = useFormContext()
+  const { findAddress } = usePostcodeLookup(apiEndpoint)
+  const { onAddressSearch, isSubmitting, error, addressList } = useAddressSearch(findAddress)
+  const { values: { postcode }, setFieldValue, validateForm } = useFormContext()
+
   const isUK = country.name === UNITED_KINGDOM
 
-  function onAddressSelect(address) {
+  function onSearchClick(e) {
```
Comment (yeahfro) [inline src/forms/elements/FieldAddress.jsx]: I actually find this easier to read:

```javascript
function onSearchClick(e) {
  e.preventDefault()
  
  return isEmpty(validateForm(['postcode'])) ? onAddressSearch(postcode) : null
}
```

File: src/forms/elements/FieldWrapper.jsx [hunk 4]
```diff
@@ -65,22 +56,51 @@ const StyledHint = styled(HintText)`
...
+    : children
+)
+
+const FieldWrapper = ({ name, label, legend, hint, error, showBorder, children }) => (
   <FormGroup>
-    <FieldInner legend={legend} error={error}>
+    <FieldInner legend={legend} error={error} showBorder={showBorder}>
       { label && <StyledLabel error={error} htmlFor={name}>{label}</StyledLabel> }
-      { hint && <StyledHint error={error}>{hint}</StyledHint> }
+      { hint && (showBorder
```
Comment (yeahfro) [inline src/forms/elements/FieldWrapper.jsx]: Would be nice to get some tests around this

File: src/forms/elements/FieldWrapper.jsx [hunk 5]
```diff
@@ -29,30 +34,16 @@ const StyledLegend = styled('legend')`
     margin-right: ${SPACING.SCALE_3};
     padding-left: ${SPACING.SCALE_2};
   `}
+  ${props => props.showBorder && `
```
Comment (yeahfro) [inline src/forms/elements/FieldWrapper.jsx]: Would be nice to get some tests around this

File: src/forms/elements/FieldWrapper.jsx [hunk 6]
```diff
@@ -4,14 +4,19 @@ import FormGroup from '@govuk-react/form-group'
...
+import { ERROR_COLOUR, GREY_2 } from 'govuk-colours'
 import { BORDER_WIDTH_FORM_ELEMENT_ERROR, SPACING } from '@govuk-react/constants'
+import Paragraph from '@govuk-react/paragraph'
 
 const StyledFieldset = styled('fieldset')`
   min-width: 0;
   margin: 0;
   padding: 0;
   border: 0;
+  ${props => props.showBorder && `
```
Comment (yeahfro) [inline src/forms/elements/FieldWrapper.jsx]: Would be nice to get some tests around this

File: src/forms/elements/FieldAddress.jsx [hunk 7]
```diff
@@ -1,31 +1,54 @@
...
+    e.preventDefault()
+    const noValidationErrors = isEmpty(validateForm(['postcode']))
+    if (noValidationErrors) {
+      return onAddressSearch(postcode)
```
Comment (yeahfro) [inline src/forms/elements/FieldAddress.jsx]: Would be good to get a test around this

File: src/forms/elements/FieldAddress.jsx [hunk 8]
```diff
@@ -1,31 +1,54 @@
 import React from 'react'
```
Comment (yeahfro) [inline src/forms/elements/FieldAddress.jsx]: A few other tests could do with being backfilled here, ie commenting out lines 54-57 the tests still pass.

File: src/forms/elements/FieldWrapper.jsx (hunk 4 above)
Comment (rafenden) [reply src/forms/elements/FieldWrapper.jsx]: Tests added.

File: src/forms/elements/FieldAddress.jsx (hunk 3 above)
Comment (rafenden) [reply src/forms/elements/FieldAddress.jsx]: Agree!

File: src/forms/elements/FieldWrapper.jsx (hunk 5 above)
Comment (rafenden) [reply src/forms/elements/FieldWrapper.jsx]: Tests added.

File: src/forms/elements/FieldWrapper.jsx (hunk 6 above)
Comment (rafenden) [reply src/forms/elements/FieldWrapper.jsx]: Tests added.

File: src/forms/elements/FieldAddress.jsx (hunk 7 above)
Comment (rafenden) [reply src/forms/elements/FieldAddress.jsx]: @paulgain is working on this in a separate PR.

File: src/forms/elements/FieldAddress.jsx (hunk 2 above)
Comment (rafenden) [reply src/forms/elements/FieldAddress.jsx]: Good point! Brace added.

File: src/address-search/usePostcodeLookup.js (hunk 1 above)
Comment (rafenden) [reply src/address-search/usePostcodeLookup.js]: True, `createAddressCount` is no longer exported.

File: src/forms/elements/FieldAddress.jsx [hunk 9]
```diff
@@ -36,40 +65,79 @@ const FieldAddress = ({
...
+              <StatusMessage>
+                Error occurred while searching for an address. Enter the address manually.
+              </StatusMessage>
+            )}
+
+            {(addressList && addressList.length > 0) && (
+              <FormGroup>
+                <Select label="Select an address" onChange={onAddressSelect}>
+                  {addressList.map(({ id, address1 }, index) => (
+                    <option key={id} value={index}>{ address1 }</option>
```
Comment (paulgain) [inline src/forms/elements/FieldAddress.jsx]: Here `id` is `undefined`. To be fair it was never set in the original implementation.

**Comment by paulgain** [comment]
**When the address search box errors it doesn't take up the entire width.**
error-width

File: src/forms/elements/FieldAddress.jsx (hunk 9 above)
Comment (rafenden) [reply src/forms/elements/FieldAddress.jsx]: Thanks, index is now used instead of ID.

**Pull request merged**